src/Feersum/Macros.fs

```diff
@@ -0,0 +1,166 @@
+module Macros
+open Syntax
+open Diagnostics
+
+// TODO: Macro patterns shold support `...` matches
+/// The macro pattern type. Used in syntax cases to define the form that a
+/// macro should match.
+type MacroPattern =
+    | Constant of SyntaxConstant
+    | Underscore
+    | Literal of string
+    | Variable of string
+    | Repeat of MacroPattern
+    | Form of MacroPattern list
+    | DottedForm of MacroPattern list * MacroPattern
+    
+type MacroBinding = (string * AstNode)
+
+let private collectResults (input: Result<'a,'b> list): Result<'a list, 'b> =
```

Comment:
Think we might want a `ResultEx` and `OptionEx` modules for these kinds of helpers.